src/functions/newsletter-submission.js

```diff
@@ -0,0 +1,51 @@
+const fetch = require("node-fetch");
+const { BUTTONDOWN_API_KEY } = process.env;
+
+const config = (userEmail) => {
+  return {
+    method: "POST",
+    headers: {
+      Authorization: `Token ${BUTTONDOWN_API_KEY}`,
+      "Content-Type": "application/json",
+    },
+    body: JSON.stringify({ email: userEmail }),
+  };
+};
+
+exports.handler = async ({ body }) => {
+  const email = getEmail(body);
+
+  try {
+    const response = await fetch(
+      "https://api.buttondown.email/v1/subscribers",
+      config(email)
+    );
+
+    const data = await response.json();
+    if (!hasError(data)) {
+      return respondWith(400, "Could not sign user up to newsletter");
+    }
+    return respondWith(200, "Thanks for signing up!");
+  } catch (error) {
+    return respondWith(500, "Could not sign user up to newsletter");
+  }
+};
+
+function getEmail(email) {
+  return decodeURIComponent(email).slice(6).split("&")[0];
```

Comment:
I think this is a bit too brittle because it's relying on that `slice(6)`. A bit too error prone. Instead, we can use `URLSearchParams` https://nodejs.org/api/url.html#url_new_urlsearchparams_string It's built in as of node.js v10.

```
const params = new URLSearchParams(body);
return params.get('email');
```

This is another situation where I would go away from this `getEmail` function and instead use `URLSearchParams` directly in the handler. Because we'll also need to get the value of `tags` from the params too.

Reply:
Wow, this is muuuch nicer than what I was doing.